feat(router): add UrlSegment[] to CanLoad interface

marcuskrahl opened the pull request:
**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


**What kind of change does this PR introduce?** (check one with "x")
```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

**What is the current behavior?** (You can also link to an open issue here)

The `Route` object as passed to implementations of `CanLoad` which only provides minimal information on the page which should be navigated to. (see #12411)

**What is the new behavior?**

Additionally to `Route` an `UrlSegment[]` is passed to implementations of `CanLoad` as a second parameter. It contains the array of path elements the user tried to navigate to before `canLoad` is evaluated.

**Does this PR introduce a breaking change?** (check one with "x")
```
[ ] Yes
[x] No
```

**Other information**:


CanLoad now defines UrlSegment[] as a second parameter of the function.
Users can store the initial url segments and refer to them later, e.g. to go
back to the original url after authentication via router.navigate(urlSegments).
Existing code still works as before because the second function parameter
does not have to be defined.

Closes #12411

jhuntoo commented:
@vicb any chance of getting this one merged soon ? It solves a valid use-case (capturing the url without having to load a lazy module), and it appears to be a non breaking change.

jasonaden commented on modules/@angular/router/src/apply_redirects.ts:
```diff
@@ -273,11 +273,12 @@ class ApplyRedirects {
   }
 
-  private getChildConfig(injector: Injector, route: Route): Observable<LoadedRouterConfig> {
+  private getChildConfig(injector: Injector, route: Route, urlSegments: UrlSegment[]):
```
For consistency with other methods receiving `UrlSegment[]`, please name the param `segments`.

jasonaden commented on modules/@angular/router/src/apply_redirects.ts:
```diff
@@ -383,15 +384,16 @@ class ApplyRedirects {
 
-function runGuards(injector: Injector, route: Route): Observable<boolean> {
+function runGuards(
+    injector: Injector, route: Route, urlSegments: UrlSegment[]): Observable<boolean> {
```
Name it `segments`.

jasonaden commented on modules/@angular/router/src/interfaces.ts:
```diff
@@ -328,10 +329,9 @@ export interface Resolve<T> {
- *     route: Route
- *   ): Observable<boolean>|Promise<boolean>|boolean {
- *     return this.permissions.canLoadChildren(this.currentUser, route);
+ *   canLoad(route: Route, urlSegments: UrlSegment[]): Observable<boolean>|Promise<boolean>|boolean
```
This formatting looks bad. Can you reformat more similarly to how it was before?

jasonaden reviewed (approved):
Looks good overall. Please make a couple minor updates, rebase, and let's get this merged. Sorry it's been sitting for a while.

jasonaden commented:
@marcuskrahl I went ahead and tried to get this rebased for you. But it's not working after the changes I made. Basically, when you submitted this PR all the files were under `modules/@angular/router` and now they are under `packages/router`. [Here is the branch](https://github.com/jasonaden/angular/tree/rebase_13127) where I rebased. If you want this commit, take it. Otherwise, can you get this rebased and sent back in?

marcuskrahl committed:
feat(router): add UrlSegment[] to CanLoad interface

CanLoad now defines UrlSegment[] as a second parameter of the function.
Users can store the initial url segments and refer to them later, e.g. to go
back to the original url after authentication via router.navigate(urlSegments).
Existing code still works as before because the second function parameter
does not have to be defined.

Closes #12411

marcuskrahl commented:
@jasonaden I made the necessary changes and everything should work as expected again. The unit tests run successfully. However the travis e2e test seems to be broken for everyone for now so I am waiting for this to be fixed.

evasquez26 commented:
What's the status of this?

MickL commented:
Please merge

picosam commented:
Any idea about when this should be merged?

marcuskrahl commented:
I rebased again. Giving this another try to get merged

digaus commented:
When will it be merged?

marcuskrahl commented:
I fixed the conflicts so this PR should be good to merge again

PoiScript commented:
Any news? I'm hoping that this can be merged and included in the v6.

ratidzidziguri commented:
Any update on this? we also need this fix to be available.
